esphome/components/pulse_counter/pulse_counter_sensor.cpp

```diff
@@ -29,17 +29,31 @@ void IRAM_ATTR BasicPulseCounterStorage::gpio_intr(BasicPulseCounterStorage *arg
     case PULSE_COUNTER_DISABLE:
       break;
     case PULSE_COUNTER_INCREMENT: {
-      auto x = arg->counter + 1;
-      arg->counter = x;
+      if (arg->dir_pin != nullptr && arg->dir_pin->digital_read()) {
+        auto x = arg->counter - 1;
+        arg->counter = x;
+      } else {
+        auto x = arg->counter + 1;
+        arg->counter = x;
+      }
```

Comment:
I'd prefer something like this:
```suggestion
      auto x = arg->counter;
      if (arg->dir_pin != nullptr && arg->dir_pin->digital_read()) {
        --x;
      } else {
        ++x;
      }
      arg->counter = x;
```

Could it then be shortened to this? How does overflow behave in this scenario? 
```suggestion
      if (arg->dir_pin != nullptr && arg->dir_pin->digital_read()) {
        --arg->counter;
      } else {
        ++arg->counter;
      }
```

This is shorter but ugly:
```cpp
      arg->counter += (arg->dir_pin != nullptr && arg->dir_pin->digital_read()) ? -1 : 1;
```